### PR TITLE
Feature/s3 browser

### PR DESCRIPTION
**Pull request by shruti0085**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)

## Description

To support S3 Browser, S3 Service node has been created. S3 Fields(bucket, object, directory) have also been mapped to a virtual file system.

## Motivation and Context

To support S3 service through AWS Toolkit for JetBrains(IntelliJ)

## Testing

S3 Service node has been tested to check if the buckets are loaded in alphabetical order.

S3ServiceNodeTest 

## Checklist

- [x ] I have read the **README** document
- [ x] I have read the **CONTRIBUTING** document
- [ x] Local run of `gradlew check` succeeds
- [ x] My code follows the code style of this project
- [ x] I have added tests to cover my changes
- [ x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    val client: S3Client
+) : S3Key(bucketName, key)
+
+open class S3Directory(
+    open val bucketName: String,
+    key: String,
+    private val client: S3Client
+) : S3Key(bucketName, key) {
+
+    fun children(): List<S3Key>? {
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt]: Is it possible for this to return an empty list instead of null? It looks like you're most of the way there.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VFS.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    override fun addVirtualFileListener(listener: VirtualFileListener) {}
+
+    override fun findFileByPath(path: String): VirtualFile? = null
+
+    override fun refreshAndFindFileByPath(path: String): VirtualFile? = null
+
+    override fun removeVirtualFileListener(listener: VirtualFileListener) {}
+
+    @Throws(UnsupportedOperationException::class)
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VFS.kt]: I think it's safe to leave out the `@Throws` annotations. They're only used for generating Java function signatures when you're [calling Kotlin code from Java](https://kotlinlang.org/docs/reference/java-to-kotlin-interop.html#checked-exceptions), and I don't see any of that here.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VFS.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    override fun renameFile(requestor: Any?, vFile: VirtualFile, newName: String) {
+        throw UnsupportedOperationException("renameFile() cannot be called against this object type")
+    }
+
+    @Throws(UnsupportedOperationException::class)
+    override fun copyFile(requestor: Any?, virtualFile: VirtualFile, newParent: VirtualFile, copyName: String): VirtualFile {
+        throw UnsupportedOperationException("copyFile() cannot be called against this object type")
+    }
+    @Throws(UnsupportedOperationException::class)
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VFS.kt]: add new line between a new method declaration and the ending curly brace of the previous one.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    open val key: S3Key
+) : VirtualFile() {
+    override fun getName(): String = key.key
+
+    override fun isWritable(): Boolean = false
+
+    override fun getPath(): String = "${key.bucket}/${key.key}"
+
+    override fun isValid(): Boolean = true
+    override fun getParent(): VirtualFile? = parent
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt]: add a new line

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    override fun getChildren(): Array<VirtualFile> =
+        directory.children()!!.sortedBy { it.name }.filterNot { it.key == directory.key }
+            .map {
+                when (it) {
+                    is S3Object -> S3VirtualFile(fileSystem, it, this)
+                    is S3Directory -> S3VirtualDirectory(fileSystem, it, this)
+                }
+            }.toTypedArray()
+
+
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt]: remove extra new line

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    val mockClientManagerRule = MockClientManagerRule(projectRule)
+
+    private val mockClient: S3Client by lazy { mockClientManagerRule.register(S3Client::class, delegateMock()) }
+
+    @Test
+    fun s3BucketsSortedAlphabetically() {
+
+        val mockSdkResponse = mock<SdkHttpResponse>()
+        mockSdkResponse.stub {
+            on { headers() } doReturn (mapOf("x-amz-bucket-region" to listOf("us-east-1")))
```
Comment (rli) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt]: I think I prefer the infix invocation of `doReturn` to remove extra parenthesis.

```kotlin
on { headers() } doReturn mapOf("x-amz-bucket-region" to listOf("us-east-1"))
```

@abrooksv There are too many ways of using mockito-kotlin. What should we standardize on?

```kotlin
`when`(something).thenReturn(something)
whenever(something).thenReturn(something)
on { something }.thenReturn(something)
on { something } doReturn something
on { something }.doReturn(something)
...
...
a bunch of other ways of doing the same thing
...
...
```

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import javax.swing.tree.DefaultMutableTreeNode
+import javax.swing.tree.DefaultTreeModel
+
+import software.aws.toolkits.jetbrains.core.credentials.ProjectAccountSettingsManager
+import software.aws.toolkits.jetbrains.core.explorer.AwsExplorerNode
+import software.aws.toolkits.jetbrains.core.explorer.AwsExplorerResourceNode
+import software.aws.toolkits.jetbrains.core.explorer.AwsExplorerServiceRootNode
+import software.aws.toolkits.jetbrains.core.explorer.AwsNodeAlwaysExpandable
+
+class S3ServiceNode(project: Project) : AwsExplorerServiceRootNode(project, "S3"),
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: You probably want to rebase off master - this has changed slightly the `AwsExplorerService` enum has a `displayName` that can be used here.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                .filter { client.regionForBucket(it.name()) == activeRegionId }
+
+        val allS3Buckets = response.map { S3Bucket(bucketName = it.name(), client = client, creationDate = it.creationDate()) }
+        return allS3Buckets
+                .sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.name })
+                .map { mapResourceToNode(it) }
+                .toList()
+    }
+
+    private fun mapResourceToNode(resource: S3Bucket) = S3BucketNode(project!!, resource)
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: Should be able to use `nodeProject` here rather than having to use the `!!` (always a sign of a problem). It looks like the `LambdaServiceNode` also uses `project!!` so if you could fix that too - that'd be good.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+class S3BucketNode(project: Project, val bucket: S3Bucket) :
+        AwsExplorerResourceNode<String>(project, S3Client.SERVICE_NAME, bucket.name, AwsIcons.Resources.CLOUDFORMATION_STACK) {
+
+    val s3Bucket: S3Bucket = bucket
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: what's the point of this assignment? can't we just use `bucket`?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    private val editorManager = FileEditorManager.getInstance(project)
+    val client: S3Client = AwsClientManager.getInstance(project).getClient()
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: I think the pattern we've used in the past is to pass the `client` instance down to child-nodes so that we're certain we're using the same credential/region and there's not a weird race-condition.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+    override fun getChildren(): Collection<AbstractTreeNode<Any>> = emptyList()
+
+    override fun resourceType(): String = "bucket"
+
+    override fun onDoubleClick(model: DefaultTreeModel, selectedElement: DefaultMutableTreeNode) {
+        val virtualBucket = S3VirtualBucket(S3VFS(client), bucket)
+        editorManager.openTextEditor(OpenFileDescriptor(project!!, virtualBucket), true)
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: again `nodeProject` here.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+sealed class S3Key(val bucket: String, val key: String) {
+    open val name = if (key.endsWith("/")) {
+        key.dropLast(1)
+    } else {
+        key.substringAfterLast("/")
+    }
+}
+
+class S3Bucket(
+    override val bucketName: String,
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt]: is there a difference between `bucketName` and `bucket` that's defined on the parent class (`S3Key`)? I don't see the need to `override` the value in each implementation.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+) : S3Key(bucketName, key) {
+
+    fun children(): List<S3Key>? {
+
+        val request = ListObjectsV2Request.builder().bucket(bucketName)
+            .delimiter("/").prefix(key).build()
+        val response = client.listObjectsV2(request)
+
+        // gives common prefixed folders
+        val folders = (response!!.commonPrefixes() ?: emptyList())
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt]: again - avoid `!!` if you can - in this case you could do `response?.commonPrefixes() ?: ...`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VFS.kt [hunk 14]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.services.s3
+
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.VirtualFileListener
+import com.intellij.openapi.vfs.VirtualFileSystem
+import software.amazon.awssdk.services.s3.S3Client
+
+class S3VFS(
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VFS.kt]: Can we expand the name `VFS` isn't a commonly understood acronym.

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import software.amazon.awssdk.services.s3.model.HeadBucketResponse
+import software.amazon.awssdk.services.s3.model.ListBucketsRequest
+import software.amazon.awssdk.services.s3.model.ListBucketsResponse
+import software.amazon.awssdk.services.s3.model.Bucket
+
+import software.aws.toolkits.jetbrains.core.MockClientManagerRule
+import software.aws.toolkits.jetbrains.utils.delegateMock
+import java.time.Instant
+
+class S3ServiceNodeTest {
```
Comment (kiiadi) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt]: Definitely needs more tests!

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt (hunk 6 above)
Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt]: I do not have a strong preference, but I agree we should standardize. Thanks mockito for all the aliases...

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+import software.aws.toolkits.jetbrains.core.explorer.AwsExplorerServiceRootNode
+import software.aws.toolkits.jetbrains.core.explorer.AwsNodeAlwaysExpandable
+
+class S3ServiceNode(project: Project) : AwsExplorerServiceRootNode(project, "S3"),
+        AwsNodeAlwaysExpandable {
+    override fun serviceName() = S3Client.SERVICE_NAME
+    private val client: S3Client = AwsClientManager.getInstance(project).getClient()
+
+    override fun loadResources(paginationToken: String?): Collection<AwsExplorerNode<*>> {
+        val activeRegionId = ProjectAccountSettingsManager.getInstance(project!!).activeRegion.id
```
[the remark this answers is not in the document]
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: That's `getProject` from the base node in the framework which is actually marked as `@Nullable`, we know it's non-null we have our own representation in `nodeProject`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java [hunk 17]
```diff
@@ -84,7 +81,7 @@ public String getId() {
         return "aws";
     }
 
-    @Nls(capitalization = Capitalization.Title)
+//    @Nls(capitalization = Capitalization.Title)
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java]: Just delete this instead of commenting

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java [hunk 18]
```diff
@@ -17,18 +15,17 @@
...
-
-import org.jetbrains.annotations.Nls;
-import org.jetbrains.annotations.Nls.Capitalization;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import software.aws.toolkits.jetbrains.services.lambda.sam.SamCommon;
 import software.aws.toolkits.jetbrains.services.telemetry.MessageBusService;
 import software.aws.toolkits.jetbrains.services.telemetry.TelemetryEnabledChangedNotifier;
 
+import javax.swing.*;
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java]: Avoid * imports

File: jetbrains-core/resources/META-INF/plugin.xml [hunk 19]
```diff
@@ -30,33 +30,26 @@
     If you come across bugs with the toolkit or have feature requests, please raise an <a
         href="https://github.com/aws/aws-toolkit-jetbrains/issues">issue</a>.
 </p>
-
```
Comment (kiiadi) [inline jetbrains-core/resources/META-INF/plugin.xml]: Nit - let's make sure we don't do just whitespace changes unless it's park of an overall reformatting effort.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+sealed class S3Key(val bucket: String, val key: String) {
+    val name: String = if (key.endsWith("/")) {
+        key.dropLast(1)
+    } else {
+        key.substringAfterLast("/")
+    }
+}
+
+class S3Bucket(
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt]: nit: these could all be on one line

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+    private val client: S3Client
+) : S3Key(bucket, key) {
+
+    fun children(): List<S3Key> {
+        val request = ListObjectsV2Request.builder()
+            .bucket(bucket)
+            .delimiter("/")
+            .prefix(key)
+            .build()
+        val response = client.listObjectsV2(request)
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt]: checkout the consumer builder pattern - you could do this as : 
```kt
val response = client.listObjectsV2 { it.bucket(bucket).delimiter("/").prefix(key) }
```

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+        val folders = (response?.commonPrefixes() ?: emptyList())
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt]: instead of the outer wrapper do:

```
response.commonPrefixes()?.map { S3Directory(bucket, it.prefix(), client) } ?: emptyList()
```

Just makes the intent a little clearer

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+            .map { S3Directory(bucket, it.prefix(), client) }
+
+        val s3Objects = (response.contents()
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt]: ditto here.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+import java.io.InputStream
+import java.io.OutputStream
+import java.time.Instant
+
+/**
+ * BaseS3VirtualFile is a base class to represent a virtual file
+ */
+
+abstract class BaseS3VirtualFile(
+    val fileSystem: S3VirtualFileSystem,
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt]: could go all one line - we're good till 140 chars - doesn't matter what @millems says.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+    override fun getTimeStamp(): Long = s3Bucket.creationDate.toEpochMilli()
+
+    fun getCreationDate(): Instant = s3Bucket.creationDate
+
+    fun getVirtualBucketName(): String = s3Bucket.bucket
+
+    override fun getChildren(): Array<VirtualFile> =
+        s3Bucket.children().sortedBy { it.bucket }
+            .map {
+                when (it) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt]: nice use of `when` matchers!

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java [hunk 26]
```diff
@@ -17,18 +15,18 @@
...
-
-import org.jetbrains.annotations.Nls;
-import org.jetbrains.annotations.Nls.Capitalization;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import software.aws.toolkits.jetbrains.services.lambda.sam.SamCommon;
 import software.aws.toolkits.jetbrains.services.telemetry.MessageBusService;
 import software.aws.toolkits.jetbrains.services.telemetry.TelemetryEnabledChangedNotifier;
 
+import javax.swing.JPanel;
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java]: Seems like your import optimization rules are different...

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/CreateS3BucketDialogTest.kt [hunk 27]
```diff
@@ -30,14 +30,14 @@ class CreateS3BucketDialogTest {
     @Test
     fun validateBucketName_emptyBucketName() {
         runInEdtAndWait {
-            val dialog = CreateS3BucketDialog(project = projectRule.project, s3Client = s3Mock)
-            val view = dialog.getViewForTesting()
-            view.bucketName.text = "  "
+        val dialog = CreateS3BucketDialog(project = projectRule.project, s3Client = s3Mock)
```
Comment (kiiadi) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/CreateS3BucketDialogTest.kt]: weird spacing - looks like the original was more correctly indented.

**Comment by kiiadi** [comment]
@shruti0085 it also might make sense to see how far the `feature/browser` branch in this repo is from your fork. We prolly wanna merge from `master` before we look to pull this in.

**Comment by kiiadi** [comment]
@shruti0085 can you rebase this from master?

**Comment by shruti0085** [comment]
@kiiadi  Sure. Can we also update feature/s3-browser in this repo by merging with master?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt [hunk 28]
```diff
@@ -106,7 +106,8 @@ class ExplorerToolWindow(private val project: Project) : SimpleToolWindowPanel(t
         object : DoubleClickListener() {
             override fun onDoubleClick(event: MouseEvent): Boolean {
                 val path = awsTree.getClosestPathForLocation(event.x, event.y)
-                (path?.lastPathComponent as? AwsExplorerNode<*>)?.onDoubleClick()
+                val selected = path?.lastPathComponent as? DefaultMutableTreeNode
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt]: Why is this needed?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+import software.aws.toolkits.jetbrains.core.explorer.nodes.AwsExplorerServiceRootNode
+import software.aws.toolkits.jetbrains.services.s3.resources.S3Resources
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+
+class S3ServiceNode(project: Project) : AwsExplorerServiceRootNode(project, AwsExplorerService.S3) {
+    private val client: S3Client = AwsClientManager.getInstance(project).getClient()
+
+    private val activeRegionId = ProjectAccountSettingsManager.getInstance(nodeProject).activeRegion.id
+    override fun getChildrenInternal(): List<AwsExplorerNode<*>> {
+        val allS3Buckets = AwsResourceCache.getInstance(nodeProject)
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: This isnt "allS3Buckets", but rather buckets in a region.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+        val virtualBucket = S3VirtualBucket(S3VirtualFileSystem(client), bucket)
+        editorManager.openTextEditor(OpenFileDescriptor(nodeProject, virtualBucket), true)
+        TelemetryService.getInstance().record(nodeProject, "s3") {
+            datum("openeditor") {
+                count()
+            }
+        }
+    }
+
+    override fun toString(): String = bucket.bucket
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: override the displayName instead.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+        return allS3Buckets.map { S3Bucket(it.name(), client, it.creationDate()) }
+            .map { S3BucketNode(nodeProject, it, client) }
+            .toList()
+    }
+}
+
+class S3BucketNode(project: Project, val bucket: S3Bucket, val client: S3Client) :
+    AwsExplorerResourceNode<String>(project, S3Client.SERVICE_NAME, bucket.bucket, AwsIcons.Resources.CLOUDFORMATION_STACK) {
+
+    private val editorManager = FileEditorManager.getInstance(project)
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ExplorerNode.kt]: Inline this into onDoubleClick()

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt [hunk 32]
```diff
@@ -0,0 +1,120 @@
...
+
+    override fun getTimeStamp(): Long = file.lastModified.toEpochMilli()
+
+    fun formatLastModified(): String {
+        val datetime = LocalDateTime.ofInstant(file.lastModified, ZoneId.systemDefault())
+        return datetime.atZone(ZoneId.systemDefault())
+            .format(DateTimeFormatter.ofPattern("MMM d YYYY hh:mm:ss a z"))
+    }
+
+    fun formatSize(): String {
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt]: Can use `com.intellij.openapi.util.text.StringUtil#formatFileSize(long)`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt [hunk 33]
```diff
@@ -0,0 +1,120 @@
...
+    }
+}
+
+open class S3VirtualBucket(fileSystem: S3VirtualFileSystem, val s3Bucket: S3Bucket) :
+    BaseS3VirtualFile(fileSystem, parent = null, key = s3Bucket) {
+
+    override fun getTimeStamp(): Long = s3Bucket.creationDate.toEpochMilli()
+
+    fun getCreationDate(): String {
+        val datetime = LocalDateTime.ofInstant(s3Bucket.creationDate, ZoneId.systemDefault())
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3VirtualFields.kt]: You have this in 2 locations, make a utility method

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/resources/S3Resources.kt [hunk 34]
```diff
@@ -0,0 +1,25 @@
...
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.S3Exception
+import software.aws.toolkits.jetbrains.core.ClientBackedCachedResource
+
+object S3Resources {
+    val LIST_BUCKETS = ClientBackedCachedResource(S3Client::class, "s3.list_buckets") {
+        listBuckets().buckets().toList()
+    }
+
+    fun regionBucket(bucketName: String) = ClientBackedCachedResource(S3Client::class, "s3.head_bucket.$bucketName") {
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/resources/S3Resources.kt]: I would name it `bucketRegion`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java [hunk 35]
```diff
@@ -17,11 +17,11 @@
 import com.intellij.ui.components.JBCheckBox;
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettingsConfigurable.java]: You can probably revert this file now. We fixed the Nls issue in master

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt [hunk 36]
```diff
@@ -21,8 +21,8 @@ const val CODEPIPELINE_SYSTEM_TAG_KEY = "aws:codepipeline:pipelineArn"
  */
 enum class TaggingResourceType(val value: String, val tagFilter: String) {
     LAMBDA_FUNCTION(message("codepipeline.lambda.resource_type"), "lambda:function"),
-    CLOUDFORMATION_STACK(message("codepipeline.stack.resource_type"), "cloudformation:stack");
-
+    CLOUDFORMATION_STACK(message("codepipeline.stack.resource_type"), "cloudformation:stack"),
+    S3_BUCKET(message("codepipeline.bucket.resource_type"), "s3:bucket");
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: is "s3:bucket" the correct type for this?

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt [hunk 37]
```diff
@@ -0,0 +1,97 @@
...
+            "abc",
+            "AEF",
+            "bcd",
+            "ZZZ"
+        )
+    }
+
+    @Test
+    fun noBucketsInTheRegion() {
+
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt]: Delete blank line

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt [hunk 38]
```diff
@@ -0,0 +1,97 @@
...
+        val bucketList = emptyList<String>()
+        resourceCache().s3buckets(bucketList)
+        bucketList.map { resourceCache().bucketRegion(it) }
+        val children = S3ServiceNode(projectRule.project).children
+        assertThat(children).allMatch { it is AwsExplorerEmptyNode }
+    }
+
+    @Test
+    fun errorLoadingBuckets() {
+        val children = S3ServiceNode(projectRule.project).children
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt]: What causes this test to throw the exception?

File: jetbrains-core/resources/META-INF/plugin.xml [hunk 39]
```diff
@@ -236,4 +236,4 @@
         </group>
```
Comment (abrooksv) [inline jetbrains-core/resources/META-INF/plugin.xml]: Revert this file

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt (hunk 36 above)
Comment (shruti0085) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/ResourceOperationAgainstCodePipeline.kt]: Fixed it to "s3"

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt (hunk 28 above)
Comment (shruti0085) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt]: `(path?.lastPathComponent as? AwsExplorerNode<*>)?.onDoubleClick()`
This doesn't call the onDoubleClick() from AwsExplorerNode. It goes to the onClick method from DoubleClickListener.

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt (hunk 38 above)
Comment (shruti0085) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt]: "IllegalStateException: No value found for ClientCachedBackedResource in mock" when it tried calling "LIST_BUCKETS"

Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt]: Change the test to be explicit in behavior. Put a failed future in the cache for example.
